sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuGenerateExec.scala

```diff
@@ -36,75 +36,219 @@ class GpuGenerateExecSparkPlanMeta(
+   * that generator is an integrated Table transformer instead of column transformer in terms of
+   * cuDF.
+   *
+   * @param inputBatch projected input data, which ensures appending columns are ahead of
+   *                   generators' inputs. So, generators can distinguish them with an offset.
+   * @param generatorOffset column offset of generator's input columns in `inputBatch`
+   * @param outer when true, each input row will be output at least once, even if the output of the
+   *              given `generator` is empty.
+   *
+   * @return result ColumnarBatch
+   */
+  def generate(inputBatch: ColumnarBatch,
+    generatorOffset: Int,
+    outer: Boolean): ColumnarBatch
+
+  /**
+   * Determine split number for generator's input batches.
+   *
+   * This is a specialized method for GPU runtime, which is called by GpuGenerateExec to split up
+   * input batches to reduce total memory cost during generating. It is necessary because most of
+   * generators may produce multiple records for each input record, which make output batch size
+   * much larger than input size.
+   *
+   * @param generatorInput input batch, only containing columns for generation
+   * @param outer when true, each input row will be output at least once, even if the output of the
+   *              given `generator` is empty.
+   *
+   * @return recommended number of splits for given input batch
+   */
+  def inputSplitNum(generatorInput: ColumnarBatch, outer: Boolean): Int = 1
+}
+
+abstract class GpuExplodeBase extends GpuUnevaluableUnaryExpression with GpuGenerator {
+
+  /** The position of an element within the collection should also be returned. */
+  def position: Boolean
+
+  // hive-compatible default alias for explode function ("col" for array, "key", "value" for map)
+  override def elementSchema: StructType = child.dataType match {
+    case ArrayType(et, containsNull) =>
+      if (position) {
+        new StructType()
+          .add("pos", IntegerType, nullable = false)
+          .add("col", et, containsNull)
+      } else {
+        new StructType()
+          .add("col", et, containsNull)
+      }
+    case MapType(kt, vt, valueContainsNull) =>
+      if (position) {
+        new StructType()
+          .add("pos", IntegerType, nullable = false)
+          .add("key", kt, nullable = false)
+          .add("value", vt, valueContainsNull)
+      } else {
+        new StructType()
+          .add("key", kt, nullable = false)
+          .add("value", vt, valueContainsNull)
+      }
+  }
+
+  // In terms of GpuExplodeBase, split number should be array size of each row. It is because
+  // we would like to keep size of output batches close to input ones in case of OOM caused by
+  // expanded output.
+  override def inputSplitNum(genIn: ColumnarBatch, outer: Boolean): Int = {
+    // only split up input batch if it is large enough
+    if (genIn.numRows() < SPLIT_BATCH_THRESHOLD) return 1
+
```

Comment:
I would prefer to see us do something more like the following in determining the splits.
This is just pseudo code.
```
val baseVectors = GpuColumnVector.extractBases(genIn)

// Get the vectors that are going to explode in size because of the array column
val columnsThatWillBeRepeated = baseVectors.slice(0, generatorOffset)
val averageRowSizeToBeRepeated = columnsThatWillBeRepeated.map(_.getDeviceMemorySize()).sum().toDouble / genIn.numRows

// Get the  average row size of the column that we are going to explode
// along with an estimate of how much the output will explode by
val (averageExplodeColumnSize, averageExplodeColumnLength) = withResource(baseVectors(generatorOffset).getChildColumnView(0)) { listDataColumn =>
    val totalCount = listDataColumn.getRowCount
    val totalSize = listDataColumn.getDeviceMemorySize
    (totalSize.toDouble/totalCount, totalCount.toDouble/genIn.numRows)
}
// The estimated size in bytes for each output row
val estimatedOutputRowSize = averageRowSizeToBeRepeated + averageExplodeColumnSize
// The maximum number of rows per split to hit the targetSize in bytes
val maxTargetOutputRowCount =  Math.min(Int.MaxValue, targetSizeBytes/estimatedOutputRowSize).toInt
// How many output rows do we think will be produced by the explode in total
val estimatedOutputRows = genIn.numRows * averageExplodeColumnLength
// how may splits will we need to keep the output size under the target size
val numSplits = ceil(estimatedOutputRows/maxTargetOutputRowCount)
```

Reply:
According to above pseudo code, I reworked the method to determine the splits. I made some simplification on the basis of pseudo code, removing the computation of average values.